File: src/main/go/leetcode/lists/topinterview/LeetCode163.go
package topinterview

import (
	"strconv"
)

func findMissingRanges(nums []int, lower int, upper int) []string {
	if len(nums) == 0 {
		if lower == upper {
			return []string{strconv.Itoa(lower)}
		}
		return []string{strconv.Itoa(lower) + "->" + strconv.Itoa(upper)}
	}
	list := make([][]int, 0)
	if lower < nums[0] {
		list = append(list, []int{lower, nums[0] - 1})
	}
	for i := 0; i < len(nums)-1; i++ {
		if nums[i]+1 < nums[i+1] {
			list = append(list, []int{nums[i] + 1, nums[i+1] - 1})
		}
	}
	if nums[len(nums)-1] < upper {
		list = append(list, []int{nums[len(nums)-1] + 1, upper})
	}
	res := make([]string, 0)
	for i := 0; i < len(list); i++ {
		if list[i][0] == list[i][1] {
			res = append(res, strconv.Itoa(list[i][0]))
		} else {
			res = append(res, strconv.Itoa(list[i][0])+"->"+strconv.Itoa(list[i][1]))
		}
	}
	return res
}

File: src/main/java/leetcode/problems/LeetCode163.java
package leetcode.problems;

import java.util.LinkedList;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/28 21:00
 */
public class LeetCode163 {

    /*
     * 给定一个排序的整数数组nums ，其中元素的范围在 闭区间 [lower, upper] 当中，
     * 返回不包含在数组中的缺失区间。
     *
     * 输入: nums = [0, 1, 3, 50, 75], lower = 0 和 upper = 99,
     * 输出: ["2", "4->49", "51->74", "76->99"]
     * */

    public static void main(String[] args) {
        System.out.println(new LeetCode163().findMissingRanges(new int[]{0, 1, 3, 50, 75}, 0, 99));
    }

    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> res = new LinkedList<>();
        int prev = lower - 1;
        int curr;
        for (int i = 0; i <= nums.length; i++) {
            if (i == nums.length) {
                curr = upper + 1;
            } else {
                curr = nums[i];
            }

            if (curr - prev > 1) {
                res.add(getRanges(prev + 1, curr - 1));
            }
            prev = curr;
        }
        return res;
    }

    private String getRanges(int from, int to) {
        return from == to ? String.valueOf(from) : from + "->" + to;
    }
}
